code/game/objects/structures/artstuff.dm

```diff
@@ -0,0 +1,399 @@
+	finalized = TRUE
+	author_ckey = user.ckey
+	generate_proper_overlay()
+	try_rename(user)
+
+/obj/item/canvas/proc/update_overlays()
+	if(!icon_generated)
+		if(used)
+			var/mutable_appearance/detail = mutable_appearance(icon,"[icon_state]wip")
+			detail.pixel_x = 1
+			detail.pixel_y = 1
+			add_overlay(detail)
+	else
+		var/mutable_appearance/detail = mutable_appearance(generated_icon)
+		detail.pixel_x = 1
+		detail.pixel_y = 1
+		add_overlay(detail)
+
+/obj/item/canvas/proc/generate_proper_overlay()
+	if(icon_generated)
+		return
+	var/png_filename = "data/paintings/temp_painting.png"
+	var/result = rustg_dmi_create_png(png_filename,"[width]","[height]",get_data_string())
+	if(result)
+		CRASH("Error generating painting png : [result]")
+	generated_icon = new(png_filename)
+	icon_generated = TRUE
+	update_icon()
+	update_overlays()
+
+/obj/item/canvas/proc/get_data_string()
+	var/list/data = list()
+	for(var/y in 1 to height)
+		for(var/x in 1 to width)
+			data += grid[x][y]
+	return data.Join("")
+
+//Todo make this element ?
+/obj/item/canvas/proc/get_paint_tool_color(obj/item/I)
+	if(!I)
+		return
+	if(istype(I, /obj/item/toy/crayon))
+		var/obj/item/toy/crayon/C = I
+		return C.colour
+	else if(istype(I, /obj/item/pen))
+		var/obj/item/pen/P = I
+		switch(P.colour)
+			if("black")
+				return "#000000"
+			if("blue")
+				return "#0000ff"
+			if("red")
+				return "#ff0000"
+		return P.colour
+	else if(istype(I, /obj/item/soap) || istype(I, /obj/item/reagent_containers/glass/rag))
+		return canvas_color
+
+/obj/item/canvas/proc/try_rename(mob/user)
+	var/new_name = stripped_input(user,"What do you want to name the painting?")
+	if(!painting_name && new_name && user.IsAdvancedToolUser())
+		painting_name = new_name
+		SStgui.update_uis(src)
+
+/obj/item/canvas/nineteenXnineteen
+	icon_state = "19x19"
+	width = 19
+	height = 19
+	ui_x = 600
+	ui_y = 600
+	pixel_x = 6
+	pixel_y = 9
+	framed_offset_x = 8
+	framed_offset_y = 9
+
+/obj/item/canvas/twentythreeXnineteen
+	icon_state = "23x19"
+	width = 23
+	height = 19
+	ui_x = 800
+	ui_y = 600
+	pixel_x = 4
+	pixel_y = 10
+	framed_offset_x = 6
+	framed_offset_y = 8
+
+/obj/item/canvas/twentythreeXtwentythree
+	icon_state = "23x23"
+	width = 23
+	height = 23
+	ui_x = 800
+	ui_y = 800
+	pixel_x = 5
+	pixel_y = 9
+	framed_offset_x = 5
+	framed_offset_y = 6
+
+/obj/item/wallframe/painting
+	name = "painting frame"
+	desc = "The perfect showcase for your favorite deathtrap memories."
+	icon = 'icons/obj/decals.dmi'
+	icon_state = "frame-empty"
+
+/obj/structure/sign/painting
+	name = "Painting"
+	desc = "Art or \"Art\"? You decide."
+	icon = 'icons/obj/decals.dmi'
+	icon_state = "frame-empty"
+	var/obj/item/canvas/C
+	var/persistence_id
+
+/obj/structure/sign/painting/Initialize(mapload, dir, building)
+	. = ..()
+	SSpersistence.painting_frames += src
+	if(dir)
+		setDir(dir)
+	if(building)
+		pixel_x = (dir & 3)? 0 : (dir == 4 ? -30 : 30)
+		pixel_y = (dir & 3)? (dir ==1 ? -30 : 30) : 0
```

Comment:
I added a helper proc for this sort of stuff recently, you should use it :smile:. It allows you to write:
```
set_pixel_offsets_from_dir(30, -30, 30, -30)
```